benchflow-experiment-manager/application/src/main/java/cloud/benchflow/experimentmanager/scheduler/ExperimentTaskScheduler.java

```diff
@@ -378,63 +332,72 @@ public void abortExperiment(String experimentID) {
 
     try {
 
-      BenchFlowExperimentState experimentState =
-          experimentModelDAO.getExperimentState(experimentID);
+      // cancel the current running task, but let it complete before
+      cancelTask(experimentID);
 
-      switch (experimentState) {
+      /*
+      When we terminate an experiment it is important that we first cancel any running task and
+      then change the state to TERMINATED otherwise the state might be overwritten.
+      */
 
-        case START:
+      // lock to ensure execution is atomic
```

Comment:
> Only one thread can execute inside a Java code block synchronized on the same monitor object.

So it is not atomic, but accessed only by one Thread at a time. Other Threads acting on other part of the code can execute.